crates/astria-sequencer-utils/src/genesis_parser.rs

```diff
@@ -0,0 +1,68 @@
+use std::{
+    fs::File,
+    path::Path,
+};
+
+use anyhow::{
+    Context as _,
+    Result,
+};
+use serde_json::{
+    to_writer_pretty,
+    Value,
+};
+use tracing::{
+    info,
+    instrument,
+};
+
+use crate::config::GenesisParserArgs;
+
+pub struct GenesisParser;
+
+impl GenesisParser {
+    #[instrument(skip_all)]
+    pub async fn propigate_data(data: GenesisParserArgs) -> Result<()> {
+        info!(
+            source_genesis_file = data.source_genesis_file.as_str(),
+            destination_genesis_file = data.destination_genesis_file.as_str(),
+            "loading genesis json data for propigation"
+        );
+        // load sequencer genesis data
+        let source_genesis_file_path = File::open(data.source_genesis_file)
+            .context("failed to open sequencer genesis file")?;
+        let source_genesis_data: Value = serde_json::from_reader(&source_genesis_file_path)
+            .context("failed deserializing sequencer genesis state from file")?;
+
+        // load cometbft genesis data
+        let destination_genesis_file_path = File::open(data.destination_genesis_file.clone())
+            .context("failed to open cometbft genesis file")?;
+        let mut destination_genesis_data: Value =
+            serde_json::from_reader(&destination_genesis_file_path)
+                .context("failed deserializing cometbft genesis state from file")?;
+
+        // merge sequencer genesis data into cometbft genesis data
+        merge_values(&mut destination_genesis_data, &source_genesis_data);
+
+        // write new state
+        let dest_file = File::create(Path::new(data.destination_genesis_file.as_str()))
+            .context("failed to open destination genesis json file")?;
+        to_writer_pretty(dest_file, &destination_genesis_data)?;
+
+        Ok(())
+    }
+}
+
+/// Merges a source JSON Value into a destination JSON Value.
+fn merge_values(a: &mut Value, b: &Value) {
```

Comment:
Slightly simplified version (no need to reborrow using `ref mut` anymore because with [match ergonomics](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html) beingi merged for a few years now all that reborrowing is done for you (so the `a` is automatically a `&mut`, and the `b` a `&`).

Also, `serde_json::Map::entry` expects a `K: Into<String>`, and since `k: &String` and `impl From<&String> for String` you don't need to clone:

```rust
fn merge_json(a: &mut Value, b: &Value) {
    match (a, b) {
        (Value::Object(a), Value::Object(b)) => {
            for (k, v) in b {
                merge_json(a.entry(k).or_insert(Value::Null), v);
            }
        }
        (a, b) => *a = b.clone(),
    }
}
```

Reply:
replaced the function with the code above but kept getting yelled at by the compiler so I've left the clone for now.